lib/onelogin/ruby-saml/response.rb

```diff
@@ -112,78 +151,327 @@ def status_message
+      # @return [Array] Array with the Issuers (REXML::Element)
+      #
+      def issuers
+        @issuers ||= begin
+          issuers = []
           node = REXML::XPath.first(document, "/p:Response/a:Issuer", { "p" => PROTOCOL, "a" => ASSERTION })
-          node ||= xpath_first_from_signed_assertion('/a:Issuer')
-          node.nil? ? nil : node.text
+          if node
+            issuers << node.text
+          end
+
+          node = REXML::XPath.first(document, "/p:Response/a:Assertion/a:Issuer", { "p" => PROTOCOL, "a" => ASSERTION })
+          if node
+            issuers << node.text          
+          end
+
+          issuers.uniq
         end
       end
 
+      # @return [String|nil] The InResponseTo attribute from the SAML Response.
+      #
+      def in_response_to
+        @in_response_to ||= begin
+          node = REXML::XPath.first(document, "/p:Response", { "p" => PROTOCOL })
+          node.nil? ? nil : node.attributes['InResponseTo']
+        end
+      end
+
+      # @return [String|nil] Destination attribute from the SAML Response
+      #
+      def destination
+        @destination ||= begin
+          node = REXML::XPath.first(document, "/p:Response", { "p" => PROTOCOL })
+          node.nil? ? nil : node.attributes['Destination']
+        end
+      end
+
+      # @return [Array] The Audience elements from the Contitions of the SAML Response.
+      #
+      def audiences
+        @audiences ||= begin
+          audiences = []
+          nodes = xpath_from_signed_assertion('/a:Conditions/a:AudienceRestriction/a:Audience')
+          nodes.each do |node|
+            unless node.nil? or node.text.empty?
+              audiences << node.text
+            end
+          end
+          audiences
+        end
+     end
+
       private
 
-      def validate(soft = true)
-        valid_saml?(document, soft)      &&
+      # Gets the expected current_url
+      # (Right now we read this url from the Assertion Consumer Service of the Settings)
+      # TODO: Calculate the real current_url and use it.
+      # @return [String] The current url
+      #
+      def current_url
+        @current_url ||= begin
+          settings.assertion_consumer_service_url
+        end
+      end
+
+      # Validates the SAML Response (calls several validation methods)
+      # If fails, the attribute errors will contains the reason for the invalidation.
+      # @param soft [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the response is invalid or not)
+      # @param request_id [String|nil] request_id The ID of the AuthNRequest sent by this SP to the IdP (if was sent any)
+      # @return [Boolean] True if the SAML Response is valid, otherwise False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate(soft = true, request_id = nil)
+        @errors = []
         validate_response_state(soft) &&
+        validate_id                   &&
```

Comment:
Minor nitpick, but all of the whitespace after the validation and before the `&&` makes it hard to understand that this is one continuous idea ... remove the whitespace, and indent the overflow lines:

```ruby
validate_response_state(soft) &&
  validate_id &&
  validate_other &&
  #...
```